maniscan.man
.TH MANISCAN 1L
''' $Id: maniscan.man,v 3.0.3.4 1993/08/25 17:04:52 ram Exp $
'''
''' $Log: maniscan.man,v $
''' Revision 3.0.3.4  1993/08/25  17:04:52  ram
''' patch12: cleanup checkin for RCS 5.6
'''
''' Revision 3.0.3.3  91/04/19  10:19:46  ram
''' patch5: added BUGS section for maniscan
''' 
''' Revision 3.0.3.2  91/04/07  18:51:20  ram
''' patch1: merged official cshar 3.0 into beta version
''' 
''' Revision 3.0.3.1  91/01/21  11:35:07  ram
''' 3.0 baseline (ram).
''' 
'''
.SH NAME
maniscan \- scan a manifest for large or binary files
.SH SYNOPSIS
.RS
.na
.ti -.5i
.B maniscan
[
.B \-e
] [
.BI \-f\| #
] [
.BI \-h\| #
] [
.BI \-i\| name
] [
.BI \-l\| name
] [
.BI \-w\| #
] [
.B \-m
] [
.BI \-o\| name
] [
.BI \-s\| #[k]
] [
.BI \-t\| text
] [
.BI \-c\| command
] [
.B \-v
]
.ad
.RE
.SH DESCRIPTION
.I Maniscan
reads an existing ``manifest'' (usually created by
.IR findsrc (1L)
and maintained by
.IR makekit (1L))
and scans it for binary files, large files, and files with long
lines or that contain non-printing characters.
Large files are automatically split and binary files are run through
.IR uuencode (1).
All problems can be recorded in a log file that should be shipped with
the other sources being prepared.
.PP
A manifest consist of some header lines, followed by a series of lines
that describe the files contained in the shell archives.
Each line looks like this;
.RS
filename\ \ \ spaces\ \ \ optional-digits\ \ \ spaces\ \ \ text
.RE
For more information on the interpretation of the file, see the
.I makekit
documentation.
.PP
The original manifest is read from the standard input.
An alternate input file may be given by using the ``\-i'' option.
The generated manifest will be sent to the standard output.
An alternate output file may be given by using the ``\-o'' option; if
the output file exists,
.I maniscan
will try to rename it with an extension of
.IR \&.BAK \&.
.PP
As it scans the manifest,
.I maniscan
can generate a log of the problems it has found so that the recipient
will know to run
.IR uudecode (1),
use
.IR cat (1)
to re-create a large file, and so on.
To create the log file, use the ``\-l'' option to specify the name
of the file; if the name is ``\-'' the log will go to standard error.
If any problems are found, the log file will be added to the new
manifest; use the ``\-e'' option to exclude it.
.PP
The log file is written such that it can often be fed directly into a
Unix shell for execution.
If a logfile is created, and the ``\-t'' option is used, then the
text specified with the ``\-t'' flag will be echoed to the user's
terminal if the logfile is executed.
This has the same meaning as the ``\-t'' option of
.IR shar (1L).
.PP
The ``\-m'' option is the same as giving the options
\&``\-iMANIFEST \-oMANIFEST \-h2 \-lPACKNOTES'' and is commonly used when
reading existing manifests.
.PP
After copying any header to the output,
.I maniscan
scans each file specified in the manifest.
If a named file appears to be a binary file (more than a third
of the first 150 bytes are not printable characters),
.I maniscan
will create a
.IR uuencode (1)'d
version, update the manifest, and print a log message.
If the original file is named
.IR foo.bar ,
then the new file will be named
.IR foo.bar.UU .
.I Maniscan
next scans for long lines or lines that have non-printable characters in
them (most commonly ASCII control characters).
The default line length is 78 characters; this can be changed by using
the ``\-w'' option; a width of zero skips this check.
It also checks for lines that end with a space or tab; to suppress this
check use the ``\-n'' option.
(Some mailers (like those on BITNET) like to strip trailing spaces, while
some editors (like GNU Emacs) seem to like to create lines with them;
use the flag as it is appropriate for you.)
Each time it finds one of these problems,
.I maniscan
will write a log entry.
.PP
If a file named in the manifest is too large,
.I maniscan
will split it up into pieces.
The default limit is 50,000 bytes; this may be changed by using
the ``\-s'' option.
If the number given with the ``\-s'' option ends with the letter ``k''
then the size is multiplied by 1024, otherwise it is taken to be the
desired maximum size in bytes.
The manifest will be updated to refer to the split pieces, and a log
entry will be written with instructions to merge the pieces.
If the original file is named
.IR foo.bar ,
then the pieces will be named
.IR foo.bar.1 ,
.IR foo.bar.2 ,
and so on.
.PP
The scanning is done in the ``right'' way so that a single large executable
named
.I a.out
would end up being shipped as
.IR a.out.UU.1 ,
.I a.out.UU.2
\&...
.IR a.out.UU.n .
In no case does
.I maniscan
edit or remove any files mentioned in a manifest (except for the manifest
itself).
.PP
If running on a system where the filenames are of limited length,
.I maniscan
will automatically truncate the generated filename to fit inside the
limit.
To help ensure portability, the ``\-f'' flag can be used on any system;
this will limit the length of the generated filenames by truncating
the base name.
For example, with ``\-f 14'' specified, the large file
.I InternalDocumentation
will be split into pieces named
.IR InternalDocu01 ,
and so on.
Note that the period is omitted.
.PP
If a file named in the manifest does not exist,
.I maniscan
can try to create it.
The ``\-c'' option should be given the name of a command to run,
with a single ``%s'' that will replace the filename.  For example,
\&``-c "sccs get %s"'' can be used to do a simple checkout of files
from SCCS.
The command will be given the full filename as named in the manifest,
and will be invoked from within the same directory that
.I maniscan
was invoked from, so a wrapper script will often be needed to checkout
items within sub-directories:
.RS
.nf
#! /bin/sh
DIR=`expr $1 : '\e(.*\e)/.*'`
if test "X${DIR}" = "X" ; then
    DIR="/"
fi
FILE=`expr $1 : '.*/\e(.*\e)'`
cd $DIR
checkout $FILE
.fi
.RE
This option may not be available on all systems.
.PP
The ``\-v'' option prints out the current version and exits.
.SH CAVEATS
The generated file names are not portable to all systems.
.PP
.SH "SEE ALSO"
makekit(1L).
.PP
.SH BUGS
.I Maniscan
will complain it cannot open a file to scan it if it finds
a zero character file.


maniscan.c
/*
**  MANISCAN
**
**  Read a manifest, looking for large files and those with binary data.
*/
#include "shar.h"
#include <signal.h>		/* for signal catching */
static char RCS[] =
	"$Id: maniscan.c,v 3.0.3.4 1993/08/25 17:04:50 ram Exp $";
/*
 * $Log: maniscan.c,v $
 * Revision 3.0.3.4  1993/08/25  17:04:50  ram
 * patch12: cleanup checkin for RCS 5.6
 *
 * Revision 3.0.3.3  91/04/19  10:18:59  ram
 * patch5: now call clean_on_sig() to exit, thus cleaning tmp files
 * patch5: missed a isspace() call to determine uu encoding
 * 
 * Revision 3.0.3.2  91/04/07  18:51:15  ram
 * patch1: merged official cshar 3.0 into beta version
 * 
 * Revision 3.0.3.1  91/01/21  11:34:51  ram
 * 3.0 baseline (ram).
 * 
 */

#ifdef	HAVE_SYSTEM
#define ARGS	"vc:ef:h:i:l:mno:s:t:w:"
#else
#define ARGS	"vef:h:i:l:mno:s:t:w:"
#endif	/* HAVE_SYSTEM */

/*
**  Global variables.
*/
static long	Size;			/* Bigger than this is bad	*/
#ifdef	HAVE_SYSTEM
static char	*Command;		/* Command to create file	*/
#endif	/* HAVE_SYSTEM */
static int	Width = SAFE_WIDTH + 1; /* Longest safe linelength	*/
static int	TrailWhite = TRUE;	/* Check for trailing spaces?	*/
static FILE	*Logfile;		/* Where to record problems	*/
static int	NameLength;		/* Limit length of filenames?	*/
static char	TempName[TEMPSIZE];	/* Temporary file we use */


/*
**	Do some cleaning (temporary files)
*/
void
tmp_clean()
{
	(void) unlink(TempName);	/* remove temporary file */
}

/*
**	Signal handler
**	We get here when a signal is caught. If TempName holds
**	a valid name, remove it.
**	Note that TempName was made global because of this routine.
*/
static sigret_t
clean_on_sig()
{
	if (strlen(TempName) > 0)
		tmp_clean();	/* clean temporary file if needed */
	
	exit(1);		/* Signals: die under abnormal conditions */
	/* NOTREACHED */
}

/*
**  Safely add an extension to a filename.
*/
static void
AddExtension(Name, Suffix, Buffer)
    char	*Name;
    char	*Suffix;
    char	*Buffer;
{
    if (NameLength == 0 || strlen(Name) + 1 + strlen(Suffix) < NameLength)
	Sprintf(Buffer, "%s.%s", Name, Suffix);
    else
	Sprintf(Buffer, "%.*s%s", NameLength - strlen(Suffix), Name, Suffix);
}


/*
**  Note that we have to run uudecode.
*/
static int
UUnote(Smudged, Name, Newname)
    int			Smudged;
    char		*Name;
    char		*Newname;
{
    if (Logfile) {
	if (!Smudged) {
	    Smudged = TRUE;
	    Fprintf(Logfile, "\n");
	}
	Fprintf(Logfile, "# Run uudecode to create \"%s\":\n", Name);
	Fprintf(Logfile, "\tuudecode %s\n", Newname);
    }
    return Smudged;
}


/*
**  Do the grunge work of checking a file.  Note that we scan stuff more
**  than once because, e.g., after it's been uuencoded and edited it
**  might be too long.
*/
static int
CheckUp(Mfile, Name, Commentary)
    register FILE	*Mfile;
    char		*Name;
    char		*Commentary;
{
    REGISTER FILE	*F;
    REGISTER FILE	*Sfile;
    REGISTER char	*p;
    REGISTER int	i;
    REGISTER int	n;
    REGISTER long	s;
    char		buff[READ_CHUNK];
    char		Newname[LINE_SIZE];
    char		Oldname[LINE_SIZE];
    char		Digits[10];
    int			Encoded;
    int			Smudged;

    /* Does file exist? */
    n = !Fexists(Name);
#ifdef	HAVE_SYSTEM
    if (n && Command) {
	/* No -- try to make it. */
	(void)sprintf(buff, Command, Name);
	(void)system(buff);
	n = !Fexists(Name);
    }
#endif	/* HAVE_SYSTEM */
    if (n) {
	Fprintf(stderr, "File \"%s\" doesn't exist.\n",
		Name);
	return FALSE;
    }

    /* For now, we do nothing about directories. */
    if (Ftype(Name) == F_DIR) {
	Fprintf(Mfile, "%s%s\n", Name, Commentary);
	return FALSE;
    }

    /* Open file, read a chunk. */
    if ((F = fopen(Name, "r")) == NULL) {
	Fprintf(stderr, "Can't open \"%s\" to check it, %s.\n",
		Name, strerror(errno));
	return FALSE;
    }
    if ((n = fread(buff, sizeof buff[0], READ_CHUNK, F)) <= 0) {
	Fprintf(stderr, "Can't read \"%s\" to check it, %s.\n",
		Name, strerror(errno));
	(void)fclose(F);
	return FALSE;
    }

    /* Is it binary?  Rough hueristic -- one third of the chunk isn't
     * printable. */
    for (Smudged = FALSE, p = buff, i = 0; p < &buff[n]; p++)
	if (!CTYPE(*p) || !(isprint(*p) || isspace(*p)))
	    i++;
    Encoded = i > n / 3;
    if (Encoded) {
	AddExtension(Name, "UU", Newname);
	uuencode(Name, Newname);
	(void)strcpy(Oldname, Name);
	(void)strcpy(Name, Newname);
	(void)fclose(F);
	if ((F = fopen(Name, "r")) == NULL) {
	    Fprintf(stderr, "Can't open \"%s\" to check it, %s.\n",
		    Name, strerror(errno));
	    return UUnote(Smudged, Oldname, Newname);
	}
    }

    /* See if the input has any long lines, or bad characters. */
    if (!Encoded) {
	rewind(F);
	for (i = 1; fgets(buff, sizeof buff, F); i++) {
	    if ((p = IDX(buff, '\n')) != NULL)
		*p = '\0';
	    if (Width && (p == NULL || p >= &buff[Width]))
		if (Logfile) {
		    if (!Smudged) {
			Smudged = TRUE;
			Fprintf(Logfile, "\n");
		    }
		    Fprintf(Logfile, "# \"%s\", line %d: line too long\n",
			    Name, i);
		}

	    for (p = buff; *p; p++)
		if (!CTYPE(*p) || !(isprint(*p) || isspace(*p)))
		    if (Logfile) {
			if (!Smudged) {
			    Smudged = TRUE;
			    Fprintf(Logfile, "\n");
			}
			Fprintf(Logfile,
			"# \"%s\", line %d: non-printable character 0%o (%s)\n",
				Name, i, *p, Seechar(*p));
		    }

	    /* Is last character whitespace?  BITNET will choke. */
	    if (TrailWhite && p > buff) {
		p--;
		if (!CTYPE(*p) || isspace(*p))
		    if (Logfile) {
			if (!Smudged) {
			    Smudged = TRUE;
			    Fprintf(Logfile, "\n");
			}
			Fprintf(Logfile,
				"# \"%s\", line %d: ends with whitespace 0%o\n",
				Name, i, *p);
		    }
	    }
	}
    }

    (void)fclose(F);

    /* Small enough to fit? */
    if (Fsize(Name) <= Size) {
	Fprintf(Mfile, "%s%s\n", Name, Commentary);
	return Encoded ? UUnote(Smudged, Oldname, Newname) : Smudged;
    }

    /* Open input. */
    if ((F = fopen(Name, "r")) == NULL) {
	Fprintf(stderr, "Can't open \"%s\" to split it, %s.\n",
		Name, strerror(errno));
	return Encoded ? UUnote(Smudged, Oldname, Newname) : Smudged;
    }

    /* Open first output, write commentary. */
    i = 0;
    Sprintf(Digits, NUMB_FORMAT, ++i);
    AddExtension(Name, Digits, Newname);
    if ((Sfile = fopen(Newname, "w")) == NULL) {
	Fprintf(stderr, "Can't open \"%s\" for writing, %s.\n",
		Newname, strerror(errno));
	(void)fclose(F);
	return Encoded ? UUnote(Smudged, Oldname, Newname) : Smudged;
    }
    Fprintf(Mfile, "%s%s (part %d)\n", Newname, Commentary, i);

    /* Read input, when current output gets too big, start a new file. */
    for (s = 0; fgets(buff, sizeof buff, F); ) {
	s += strlen(buff);
	if (s > Size) {
	    /* Start a new file. */
	    (void)fclose(Sfile);
	    Sprintf(Digits, NUMB_FORMAT, ++i);
	    AddExtension(Name, Digits, Newname);
	    if ((Sfile = fopen(Newname, "w")) == NULL) {
		Fprintf(stderr, "Can't open \"%s\" for writing, %s.\n",
			Newname, strerror(errno));
		(void)fclose(F);
		return Encoded ? UUnote(Smudged, Oldname, Newname) : Smudged;
	    }
	    Fprintf(Mfile, "%s%s (part %d)\n", Newname, Commentary, i);
	    s = strlen(buff);
	}
	(void)fputs(buff, Sfile);
    }

    /* Close input and current output, remove temporary if we made one. */
    (void)fclose(F);
    (void)fclose(Sfile);
    if (Encoded)
	(void)unlink(Name);

    /* Write summary message. */
    if (Logfile) {
	if (!Smudged) {
	    Smudged = TRUE;
	    Fprintf(Logfile, "\n");
	}
	Fprintf(Logfile,
	    "# \"%s\" was split into %d parts; to create it, do\n",
	    Name, i);
	if (i < 10)
	    Fprintf(Logfile, "\tcat %s.0[1-9] >%s\n", Name, Name);
	else if (i < 100)
	    Fprintf(Logfile, "\tcat %s.0[1-9] %s.[1-9][0-9] >%s\n",
		    Name, Name, Name);
	else
	    Fprintf(Logfile, "# Whatever is appropriate\n");
	/* ram:
	 * If file is executable, add a chmod in the log file.
	 */
	if (Fexecute(Name))
	    Fprintf(Logfile, "\tchmod +x %s\n", Name);
    }
    return Encoded ? UUnote(Smudged, Oldname, Newname) : Smudged;
}


int
main(ac, av)
    REGISTER int	ac;
    REGISTER char	*av[];
{
    REGISTER FILE	*F;
    REGISTER char	*p;
    REGISTER int	i;
    REGISTER int	Header;
    REGISTER int	LogDirty;
    char		*InName;
    char		*OutName;
    char		*LogName;
    char		*NameStart;
    char		*Trailer;
    char		buff[BUFSIZ];
    char		Rest[LINE_SIZE];
    char		Name[LINE_SIZE];
    int			ExcludeIt;
    int			Oops;

	/* ram:
	 * First initialize the temporary file name to be NULL,
	 * because when a signal SIGHUP, SIGINT SIGQUIT or
	 * SIGTERM occurs, we go to the signal handler that will
	 * try to remove any temporary file created. A NULL name
	 * (i.e. length is 0) will be ignored.
	 */
	TempName[0] = '\0';		/* NULL name */

	/* ram:
	 * Define the signal we want to trap. If the program
	 * dies due to one of these, we first check for remaining
	 * temporary files by calling clean_on_sig().
	 */
	signal(SIGHUP, clean_on_sig);
	signal(SIGINT, clean_on_sig);
	signal(SIGQUIT, clean_on_sig);
	signal(SIGTERM, clean_on_sig);
	signal(SIGBUS, clean_on_sig);	/* should never happen ! */
	signal(SIGSEGV, clean_on_sig);	/* but man never knows... */

    /* Parse JCL. */
    ExcludeIt = FALSE;
    InName = NULL;
    OutName = NULL;
    LogName = NULL;
    Header = 0;
    Size = 50000;
    Trailer = NULL;
    for (Oops = FALSE; (i = getopt(ac, av, ARGS)) != EOF; )
	switch (i) {
	default:
	    Oops = TRUE;
	    break;
	case 'v':		/* Print version			*/
	    Version(RCS);
	    /* NOTREACHED */
#ifdef	HAVE_SYSTEM
	case 'c':		/* Command to run if file doesn't exist	*/
	    Command = optarg;
	    break;
#endif	/* HAVE_SYSTEM */
	case 'e':		/* Don't include manifest in manifest	*/
	    ExcludeIt = TRUE;
	    break;
	case 'f':		/* Limit length of generated filenames	*/
	    NameLength = atoi(optarg);
	    break;
	case 'h':		/* Lines of header to skip		*/
	    Header = atoi(optarg);
	    break;
	case 'i':		/* Name of input manifest		*/
	    InName = optarg;
	    break;
	case 'l':		/* Name of packnotes log file		*/
	    LogName = optarg;
	    break;
	case 'm':		/* Convenient option shorthand		*/
	    LogName = "PACKNOTES";
	    InName = OutName = "MANIFEST";
	    Header = 2;
	    break;
	case 'n':		/* Don't check for trailing whitespace	*/
	    TrailWhite = FALSE;
	    break;
	case 'o':		/* Name for generated manifest file	*/
	    OutName = optarg;
	    break;
	case 's':		/* Maximum file size			*/
	    Size = (long)atoi(optarg);
	    if (IDX(optarg, 'k') || IDX(optarg, 'K'))
		Size *= 1024;
	    break;
	case 't':		/* Final note after unpacking		*/
	    Trailer = optarg;
	    break;
	case 'w':		/* Maximum line width			*/
	    if ((Width = atoi(optarg)) > READ_CHUNK) {
		Fprintf(stderr,
		    "Can't check widths greater than %d, option ignored.\n",
		    READ_CHUNK);
		Width = 0;
	    }
	    break;
	}
    ac -= optind;
    av += optind;

    if (ac != 0 || Oops) {
	Fprintf(stderr, "Usage:\n  maniscan %s          %s\n",
#ifdef	HAVE_SYSTEM
		"[-e] [-n] [-s#[k]] [-ttext] [-w#]",
#else
		"[-e] [-n] [-s#[k]] [-ttext] [-w#] [-c command]",
#endif	/* HAVE_SYSTEM */
		"[-m | -iMANIFEST -oMANIFEST -h2 -lPACKNOTES]");
	clean_on_sig();
	/* NOTREACHED */
    }

    /* Open the input file. */
    if (InName && freopen(InName, "r", stdin) == NULL) {
	Fprintf(stderr, "Can't read %s as manifest, %s.\n",
		InName, strerror(errno));
	clean_on_sig();
	/* NOTREACHED */
    }

    /* Open the output file. */
    if (OutName == NULL)
	F = stdout;
    else {
	MakeTempName(TempName, TEMP_NAME1);
	if ((F = fopen(TempName, "w")) == NULL) {
	    Fprintf(stderr, "Can't open '%s' for output, %s.\n",
		    TempName, strerror(errno));
	    clean_on_sig();
	    /* NOTREACHED */
	}
    }

    /* Open the log file. */
    LogDirty = FALSE;
    if (LogName) {
	if (EQ(LogName, "-")) {
	    Logfile = stderr;
	    ExcludeIt = TRUE;
	}
	else if ((Logfile = fopen(LogName, "w")) == NULL) {
	    Fprintf(stderr, "Can't open \"%s\" for output, %s.\n",
		    LogName, strerror(errno));
	    clean_on_sig();
	    /* NOTREACHED */
	}
    }

    /* Reduce size by slop so last line doesn't make us too long. */
    Size -= 2 * Width;

    /* Skip any possible prolog, then output rest of file. */
    while (--Header >= 0 && fgets(buff, sizeof buff, stdin))
	(void)fputs(buff, F);
    if (feof(stdin)) {
	Fprintf(stderr, "Nothing but header lines in the manifest.\n");
	clean_on_sig();
	/* NOTREACHED */
    }

    /* Scan rest of file. */
    while (fgets(buff, sizeof buff, stdin)) {
	if ((p = IDX(buff, '\n')) == NULL)
	    Fprintf(stderr, "Warning, line too long:\n\t%s\n", buff);
	else
	    *p = '\0';

	/* Skip leading whitespace and ignore all-blank lines. */
	for (p = buff; *p && CTYPE(*p) && isspace(*p); )
	    p++;
	if (*p == '\0')
	    continue;
	/* Save name, find end of it, stuff rest of line away. */
	for (NameStart = p; *++p && CTYPE(*p) && !isspace(*p); )
	    continue;
	i = *p;
	*p = '\0';
	(void)strcpy(Name, NameStart);
	*p = i;
	(void)strcpy(Rest, p);

	/* ram:
	 * If LogName is PACKNOTES and PACKNOTES already is in the
	 * MANIFEST, then we must not try to read it !!
	 */
	if (!(EQ(LogName, "PACKNOTES") && EQ (Name, "PACKNOTES"))
	&& CheckUp(F, Name, Rest))
	    LogDirty = TRUE;
    }

    /* Close logfile, possibly add it to the manifest. */
    if (Logfile) {
	if (Trailer && *Trailer) {
	    Fprintf(F, "# Follow these instructions:\n");
	    Fprintf(F, "\techo \"%s\"\n", Trailer);
	}
	if (Logfile != stderr)
	    (void)fclose(Logfile);
	if (LogDirty && !ExcludeIt)
	    Fprintf(F, MANI_FORMAT,
		    LogName, 1, "Warnings about long lines, etc");
    }

    /* Move temp file to new file? */
    if (OutName) {
	/* Open new file. */
	(void)fclose(F);
	if ((F = fopen(TempName, "r")) == NULL) {
	    Fprintf(stderr, "Can't open \"%s\" for reading, %s.\n",
		    Name, strerror(errno));
	    clean_on_sig();
	    /* NOTREACHED */
	}

	/* Save old manifest. */
	if (Fexists(OutName))
	    SafeRename(OutName);
	if (freopen(OutName, "w", stdout) == NULL) {
	    Fprintf(stderr, "Can't open \"%s\" for writing, %s.\n",
		    OutName, strerror(errno));
	    clean_on_sig();
	    /* NOTREACHED */
	}

	/* Copy. */
	while (fgets(buff, sizeof buff, F))
	    (void)fputs(buff, stdout);
    }

	tmp_clean();	/* remove temporary file */

    /* That's all she wrote. */
    exit(0);
    /* NOTREACHED */
}
